beacon-chain/types/active_state.go

```diff
@@ -0,0 +1,214 @@
+		if err != nil {
+			return nil, err
+		}
+
+		for _, h := range parentHashes {
+			// Skip calculating for this hash if the hash is part of oblique parent hashes.
+			var skip bool
+			for _, oblique := range attestation.ObliqueParentHashes {
+				if bytes.Equal(h[:], oblique) {
+					skip = true
+				}
+			}
+			if skip {
+				continue
+			}
+
+			// Initialize vote cache of a given block hash if it doesn't exist already.
+			if !a.isVoteCacheEmpty(h) {
+				new[h] = newVoteCache()
+			}
+
+			// Loop through attester indices, if the attester has voted but was not accounted for
+			// in the cache, then we add attester's index and balance to the block cache.
+			for i, attesterIndex := range attesterIndices {
+				var attesterExists bool
+				if !utils.CheckBit(attestation.AttesterBitfield, i) {
+					continue
+				}
+				for _, indexInCache := range new[h].VoterIndices {
+					if attesterIndex == indexInCache {
+						attesterExists = true
+					}
+				}
+				if !attesterExists {
+					new[h].VoterIndices = append(new[h].VoterIndices, attesterIndex)
+					new[h].VoteTotalDeposit += cState.Validators()[attesterIndex].Balance
+				}
+			}
+		}
+	}
+
+	return new, nil
+}
+
+// CalculateNewActiveState returns the active state for `block` based on its own state.
+// This method should not modify its own state.
+func (a *ActiveState) CalculateNewActiveState(block *Block, cState *CrystallizedState, parentSlot uint64) (*ActiveState, error) {
+	// Derive the new set of pending attestations
+	newPendingAttestations := a.calculateNewAttestations(block.data.Attestations, cState.LastStateRecalc())
+
+	// Derive the new value for RecentBlockHashes
+	newRecentBlockHashes, err := a.calculateNewBlockHashes(block, parentSlot)
+	if err != nil {
+		return nil, fmt.Errorf("failed to update recent block hashes: %v", err)
+	}
+
+	// With a valid beacon block, we can compute its attestations and store its votes/deposits in cache.
+	newBlockVoteCache, err := a.calculateNewVoteCache(block, cState)
+	if err != nil {
+		return nil, fmt.Errorf("failed to update vote cache: %v", err)
+	}
+
+	return NewActiveState(&pb.ActiveState{
+		PendingAttestations: newPendingAttestations,
+		RecentBlockHashes:   newRecentBlockHashes,
+	}, newBlockVoteCache), nil
+}
+
+// getSignedParentHashes returns all the parent hashes stored in active state up to last cycle length.
+func (a *ActiveState) getSignedParentHashes(block *Block, attestation *pb.AttestationRecord) [][32]byte {
+	var signedParentHashes [][32]byte
+	start := block.SlotNumber() - attestation.Slot
+	end := block.SlotNumber() - attestation.Slot - uint64(len(attestation.ObliqueParentHashes)) + params.CycleLength
+
+	recentBlockHashes := a.RecentBlockHashes()
+	signedParentHashes = append(signedParentHashes, recentBlockHashes[start:end]...)
+
+	for _, obliqueParentHashes := range attestation.ObliqueParentHashes {
+		hashes := common.BytesToHash(obliqueParentHashes)
```

Comment:
why are these keccak256? shouldn't we be using blake here?

Reply:
I'm not sure what you're asking here. These are already hashed, this line is just converting `[]byte` into `[32]byte`.

Reply:
Oh I see, so common.BytesToHash doesn't actually apply keccak to the slice.